en/guide/message_signing.md

```diff
@@ -1,121 +1,144 @@
+<!-- 
+How are we handling case of signed packets that meet the criteria for accepting unsigned packets but signature is incorrect or the timestamp rules are not met. E.g. for lost copter finding or re-keying? AFAIK no support in the C library for handling this?
+-->
 
-The solution adopted for MAVProxy is shown below:
 
-```
-if (msg.get_signed() and
-	self.mav.signing.link_id == 0 and
-	msg.get_link_id() != 0 and
-	self.target_system == msg.get_srcSystem() and
-	self.target_component == msg.get_srcComponent()):
-	# change to link_id from incoming packet
-	self.mav.signing.link_id = msg.get_link_id()
-```
+## Secret Key Setup {#secret_key}
+
+The secret key is 32 bytes of binary data stored on both ends of a MAVLink channel (i.e. an autopilot and a ground station or MAVLink API) and used in the calculation of the message signature.
+
+The secret key should be stored in persistent storage, and should not be exposed via any publicly accessible communication protocol, or be saved in log files. 
+In particular, the key must not be exposed in MAVLink parameters, MAVLink log files or dataflash log files that may be used for public log analysis.
+
+The method of generating the 32 byte secret key is up to the GCS implementation (or whatever system is chosen for this purpose). 
+For example, it could be generated by:
+* A user-entered string of up to 64 bytes length that is then run through SHA-256.
+* A random key generator.
+
+The secret key may be shared to target devices using the [SETUP_SIGNING](../messages/common.md#SETUP_SIGNING) message. 
+This should be done over a secure link like USB or wired Ethernet (in theory the secret might be shared over any link if the rules for accepting unsigned packets are met, but it should not be broadcast over a radio interface).
+The receiving system must be set up to process these messages and store the received secret key to the appropriate permanent storage.
+
+Autopilots that don't offer MAVLink over USB might create a module that can set the secret key from a command line interface (e.g. the NSH Shell).
+
+> **Tip** We recommend that all GCS implementations should generate the secret key from a sha256 hash of a user-provided passphrase, and share this with connected systems over a secure link. <!-- really? Random number would be safer, and a warning that it needs to be updated/changed -->
```

Comment:
Do we really recommend GCS takes string from user. A random key would be safer and easier in GCS.

Reply:
You may need to be able to setup multiple GCS's to the same key, which at somepoint means transfering it between them. Allowing the GCS to consistently hash the key is a compromise to make this work easily for users. But yes, it does have all the normal "password123" problems everywhere, and it shouldn't be the default recommendation we push at people.

Reply:
Great. I have removed the recommendation. It now just says:

> We recommend that GCS implementations should generate the secret key and share this with connected systems over a secure link.